File: src/api/diagnostics.ts
import { invoke } from "@tauri-apps/api/core";

export type FrontendErrorKind =
  | "window"
  | "unhandled_rejection"
  | "react_uncaught"
  | "react_caught"
  | "react_recoverable";

export interface FrontendErrorReportRequest {
  kind: FrontendErrorKind;
  message: string;
  stack?: string;
  componentStack?: string;
  source?: string;
  line?: number;
  column?: number;
}

export async function frontendErrorReport(request: FrontendErrorReportRequest): Promise<void> {
  await invoke("diagnostics_frontend_error_report", { request });
}

File: src-tauri/src/protocol/diagnostics.rs
use serde::Deserialize;

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FrontendErrorKind {
    Window,
    UnhandledRejection,
    ReactUncaught,
    ReactCaught,
    ReactRecoverable,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FrontendErrorReportRequest {
    pub kind: FrontendErrorKind,
    pub message: String,
    pub stack: Option<String>,
    pub component_stack: Option<String>,
    pub source: Option<String>,
    pub line: Option<u32>,
    pub column: Option<u32>,
}

impl FrontendErrorKind {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Window => "window",
            Self::UnhandledRejection => "unhandled_rejection",
            Self::ReactUncaught => "react_uncaught",
            Self::ReactCaught => "react_caught",
            Self::ReactRecoverable => "react_recoverable",
        }
    }

    pub const fn is_recoverable(self) -> bool {
        matches!(self, Self::ReactCaught | Self::ReactRecoverable)
    }
}
